### ISSUE TITLE
Error when calling REST or GRPC API from python clients

### ISSUE THREAD

**Issue by magallardo** [openvinotoolkit/model_server#384]
Hello,

I have started using the latest docker image of model-server and running into several issues and would appreciate any help to get this working.

I am running the docker image on a Mac and the docker image starts fine.  Following the output from the console:

```
docker run --user $(id -u):$(id -g) -v ~/models/server/face-detection:/models -p 9000:9000 -p 9001:9001 openvino/model_server:latest --model_path /models --model_name face-detection --port 9000 --rest_port 9001 --log_level DEBUG
[2020-12-01 20:06:44.764] [serving] [debug] CLI parameters passed to ovms server
[2020-12-01 20:06:44.764] [serving] [debug] model_path: /models
[2020-12-01 20:06:44.764] [serving] [debug] model_name: face-detection
[2020-12-01 20:06:44.764] [serving] [debug] batch_size: 0
[2020-12-01 20:06:44.764] [serving] [debug] shape: 
[2020-12-01 20:06:44.764] [serving] [debug] model_version_policy: 
[2020-12-01 20:06:44.764] [serving] [debug] nireq: 0
[2020-12-01 20:06:44.764] [serving] [debug] target_device: CPU
[2020-12-01 20:06:44.764] [serving] [debug] plugin_config: 
[2020-12-01 20:06:44.764] [serving] [debug] gRPC port: 9000
[2020-12-01 20:06:44.764] [serving] [debug] REST port: 9001
[2020-12-01 20:06:44.764] [serving] [debug] REST workers: 24
[2020-12-01 20:06:44.764] [serving] [debug] gRPC workers: 1
[2020-12-01 20:06:44.764] [serving] [debug] gRPC channel arguments: 
[2020-12-01 20:06:44.764] [serving] [debug] log level: DEBUG
[2020-12-01 20:06:44.764] [serving] [debug] log path: 
[2020-12-01 20:06:44.764] [serving] [debug] Batch size set: false, shape set: false
[2020-12-01 20:06:44.766] [serving] [debug] Currently registered versions count:0
[2020-12-01 20:06:44.766] [serving] [info] Getting model from /models
[2020-12-01 20:06:44.766] [serving] [info] Model downloaded to /models
[2020-12-01 20:06:44.766] [serving] [info] Will add model: face-detection; version: 1 ...
[2020-12-01 20:06:44.767] [serving] [info] Loading model: face-detection, version: 1, from path: /models/1, with target device: CPU ...
[2020-12-01 20:06:44.768] [serving] [info] STATUS CHANGE: Version 1 of model face-detection status change. New status: ( "state": "START", "error_code": "OK" )
[2020-12-01 20:06:44.768] [serving] [debug] setLoading: face-detection - 1 (previous state: START) -> error: OK
[2020-12-01 20:06:44.768] [serving] [info] STATUS CHANGE: Version 1 of model face-detection status change. New status: ( "state": "LOADING", "error_code": "OK" )
[2020-12-01 20:06:44.768] [serving] [debug] Getting model files from path:/models/1
[2020-12-01 20:06:44.770] [serving] [debug] Try reading model file:/models/1/face-detection-retail-0004.xml
[2020-12-01 20:06:44.857] [serving] [debug] Network shape - (1,3,300,300); Final shape - (1,3,300,300)
[2020-12-01 20:06:44.857] [serving] [info] Input name: data; mapping_name: ; shape: 1 3 300 300 ; precision: FP32, layout:NCHW
[2020-12-01 20:06:44.857] [serving] [debug] model: face-detection, version: 1; reshaping inputs is not required
[2020-12-01 20:06:44.857] [serving] [info] Output name: detection_out ; mapping name: ; shape: 1 1 200 7  ; precision: FP32, layout:NCHW
```

However, when I try to use the sample REST and gRPC python clients, I am getting errors and the clients can't connect to the server.  Following is the output for the REST client to get model status:

```
python rest_get_model_status.py --rest_port 9001 --model_name face-detection --model_version 1
Traceback (most recent call last):
  File "/Users/mg/virtualenvs/openvino/venv/lib/python3.7/site-packages/urllib3/connectionpool.py", line 706, in urlopen
    chunked=chunked,
  File "/Users/mg/virtualenvs/openvino/venv/lib/python3.7/site-packages/urllib3/connectionpool.py", line 445, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/Users/mg/virtualenvs/openvino/venv/lib/python3.7/site-packages/urllib3/connectionpool.py", line 440, in _make_request
    httplib_response = conn.getresponse()
  File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/http/client.py", line 1321, in getresponse
    response.begin()
  File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/http/client.py", line 296, in begin
    version, status, reason = self._read_status()
  File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/http/client.py", line 265, in _read_status
    raise RemoteDisconnected("Remote end closed connection without"
http.client.RemoteDisconnected: Remote end closed connection without response

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/mg/virtualenvs/openvino/venv/lib/python3.7/site-packages/requests/adapters.py", line 449, in send
    timeout=timeout
  File "/Users/mg/virtualenvs/openvino/venv/lib/python3.7/site-packages/urllib3/connectionpool.py", line 756, in urlopen
    method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]
  File "/Users/mg/virtualenvs/openvino/venv/lib/python3.7/site-packages/urllib3/util/retry.py", line 531, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/Users/mg/virtualenvs/openvino/venv/lib/python3.7/site-packages/urllib3/packages/six.py", line 734, in reraise
    raise value.with_traceback(tb)
  File "/Users/mg/virtualenvs/openvino/venv/lib/python3.7/site-packages/urllib3/connectionpool.py", line 706, in urlopen
    chunked=chunked,
  File "/Users/mg/virtualenvs/openvino/venv/lib/python3.7/site-packages/urllib3/connectionpool.py", line 445, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/Users/mg/virtualenvs/openvino/venv/lib/python3.7/site-packages/urllib3/connectionpool.py", line 440, in _make_request
    httplib_response = conn.getresponse()
  File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/http/client.py", line 1321, in getresponse
    response.begin()
  File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/http/client.py", line 296, in begin
    version, status, reason = self._read_status()
  File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/http/client.py", line 265, in _read_status
    raise RemoteDisconnected("Remote end closed connection without"
urllib3.exceptions.ProtocolError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "rest_get_model_status.py", line 35, in <module>
    result = requests.get("{}:{}/v1/models/{}{}".format(args['rest_url'], args['rest_port'], args['model_name'], version))
  File "/Users/mg/virtualenvs/openvino/venv/lib/python3.7/site-packages/requests/api.py", line 76, in get
    return request('get', url, params=params, **kwargs)
  File "/Users/mg/virtualenvs/openvino/venv/lib/python3.7/site-packages/requests/api.py", line 61, in request
    return session.request(method=method, url=url, **kwargs)
  File "/Users/mg/virtualenvs/openvino/venv/lib/python3.7/site-packages/requests/sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "/Users/mg/virtualenvs/openvino/venv/lib/python3.7/site-packages/requests/sessions.py", line 655, in send
    r = adapter.send(request, **kwargs)
  File "/Users/mg/virtualenvs/openvino/venv/lib/python3.7/site-packages/requests/adapters.py", line 498, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
```

Please advise.

Marcelo

**Comment by dtrawins** [comment]
@magallardo  I assume the logs are truncated because they should also include a note, that the gRPC and REST servers are started in OVMS container.
It is likely that the problem is caused the the proxy settings. By default, the clients are connecting to `localhost` address so make sure it is added to `no_proxy` variable or http_proxy is unset. You could just test it with `curl` command to confirm connectivity with the docker container endpoint.

**Comment by magallardo** [comment]
@dtrawins The logs are not truncated.  That is all the is displayed in the console.  I have also tried calling the service from postman and curl and got a "Could get response" error.
I will investigate using the proxy settings but for any other containers with REST endpoints on this machine I have not had to set anything.

If the logs are not complete, does it mean the REST and gRPC components are not starting?

Thanks,
Marcelo

**Comment by dtrawins** [comment]
@magallardo In the logs there should be info about started grpc and rest servers. It looks like they were not. It might be the best to check the docker container status with `docker inspect <container_name>`. Maybe your computer is running out of memory so docker engine terminated the container prematurely? Do you have some RAM restrictions for docker on the MAC?

**Comment by magallardo** [comment]
@dtrawins This machine has plenty of resources and I have been running other containers without any issue.  I have also run the Openvino Workbench container without a problem.  I also just rebooted the machine to make sure nothing else is running while I try to run the model server and the logs still don't show any note regarding REST and gRPC starting.
I also tried docker inspect and everything seems ok.  I am attaching the output from docker inspect and maybe you can help me see if something is wrong.

```
docker inspect 30aa241b7865
[
    {
        "Id": "30aa241b78656cd3bdbdef9f3e41e9917e6ffd02309535cc333eda28ba9dac5e",
        "Created": "2020-12-03T14:25:54.07138091Z",
        "Path": "/ovms/bin/ovms",
        "Args": [
            "--model_path",
            "/models/face-detection",
            "--model_name",
            "face-detection",
            "--port",
            "9000",
            "--rest_port",
            "9001",
            "--log_level",
            "DEBUG"
        ],
        "State": {
            "Status": "running",
            "Running": true,
            "Paused": false,
            "Restarting": false,
            "OOMKilled": false,
            "Dead": false,
            "Pid": 1941,
            "ExitCode": 0,
            "Error": "",
            "StartedAt": "2020-12-03T14:25:54.544586406Z",
            "FinishedAt": "0001-01-01T00:00:00Z"
        },
        "Image": "sha256:160545972851614e305926c5db97f4acc7cfac3fff0a257d745818cb476cadd0",
        "ResolvConfPath": "/var/lib/docker/containers/30aa241b78656cd3bdbdef9f3e41e9917e6ffd02309535cc333eda28ba9dac5e/resolv.conf",
        "HostnamePath": "/var/lib/docker/containers/30aa241b78656cd3bdbdef9f3e41e9917e6ffd02309535cc333eda28ba9dac5e/hostname",
        "HostsPath": "/var/lib/docker/containers/30aa241b78656cd3bdbdef9f3e41e9917e6ffd02309535cc333eda28ba9dac5e/hosts",
        "LogPath": "/var/lib/docker/containers/30aa241b78656cd3bdbdef9f3e41e9917e6ffd02309535cc333eda28ba9dac5e/30aa241b78656cd3bdbdef9f3e41e9917e6ffd02309535cc333eda28ba9dac5e-json.log",
        "Name": "/wizardly_hertz",
        "RestartCount": 0,
        "Driver": "overlay2",
        "Platform": "linux",
        "MountLabel": "",
        "ProcessLabel": "",
        "AppArmorProfile": "",
        "ExecIDs": null,
        "HostConfig": {
            "Binds": [
                "/users/mag//models/server:/models"
            ],
            "ContainerIDFile": "",
            "LogConfig": {
                "Type": "json-file",
                "Config": {}
            },
            "NetworkMode": "default",
            "PortBindings": {
                "9000/tcp": [
                    {
                        "HostIp": "",
                        "HostPort": "9000"
                    }
                ],
                "9001/tcp": [
                    {
                        "HostIp": "",
                        "HostPort": "9001"
                    }
                ]
            },
            "RestartPolicy": {
                "Name": "no",
                "MaximumRetryCount": 0
            },
            "AutoRemove": false,
            "VolumeDriver": "",
            "VolumesFrom": null,
            "CapAdd": null,
            "CapDrop": null,
            "Capabilities": null,
            "Dns": [],
            "DnsOptions": [],
            "DnsSearch": [],
            "ExtraHosts": null,
            "GroupAdd": null,
            "IpcMode": "private",
            "Cgroup": "",
            "Links": null,
            "OomScoreAdj": 0,
            "PidMode": "",
            "Privileged": false,
            "PublishAllPorts": false,
            "ReadonlyRootfs": false,
            "SecurityOpt": null,
            "UTSMode": "",
            "UsernsMode": "",
            "ShmSize": 67108864,
            "Runtime": "runc",
            "ConsoleSize": [
                0,
                0
            ],
            "Isolation": "",
            "CpuShares": 0,
            "Memory": 0,
            "NanoCpus": 0,
            "CgroupParent": "",
            "BlkioWeight": 0,
            "BlkioWeightDevice": [],
            "BlkioDeviceReadBps": null,
            "BlkioDeviceWriteBps": null,
            "BlkioDeviceReadIOps": null,
            "BlkioDeviceWriteIOps": null,
            "CpuPeriod": 0,
            "CpuQuota": 0,
            "CpuRealtimePeriod": 0,
            "CpuRealtimeRuntime": 0,
            "CpusetCpus": "",
            "CpusetMems": "",
            "Devices": [],
            "DeviceCgroupRules": null,
            "DeviceRequests": null,
            "KernelMemory": 0,
            "KernelMemoryTCP": 0,
            "MemoryReservation": 0,
            "MemorySwap": 0,
            "MemorySwappiness": null,
            "OomKillDisable": false,
            "PidsLimit": null,
            "Ulimits": null,
            "CpuCount": 0,
            "CpuPercent": 0,
            "IOMaximumIOps": 0,
            "IOMaximumBandwidth": 0,
            "MaskedPaths": [
                "/proc/asound",
                "/proc/acpi",
                "/proc/kcore",
                "/proc/keys",
                "/proc/latency_stats",
                "/proc/timer_list",
                "/proc/timer_stats",
                "/proc/sched_debug",
                "/proc/scsi",
                "/sys/firmware"
            ],
            "ReadonlyPaths": [
                "/proc/bus",
                "/proc/fs",
                "/proc/irq",
                "/proc/sys",
                "/proc/sysrq-trigger"
            ]
        },
        "GraphDriver": {
            "Data": {
                "LowerDir": "/var/lib/docker/overlay2/958ae73a3ae371bcd406adcbaf5082bfaf9a750f75b451f1afe78de0d00ed8c2-init/diff:/var/lib/docker/overlay2/09ba465c914394b48f0fd98ba6c12e14ef9d1d87c206cc0cfb2e675ed99c4dba/diff:/var/lib/docker/overlay2/f4b703663cea41519cad7a6a8b642bc22b8f56c5bfc55d45d8f94348ec9beb9c/diff:/var/lib/docker/overlay2/05192d889e2bc18af2b8b92d450b6d4996e5a1f2d0e5e00ff144a07aea8cd05f/diff",
                "MergedDir": "/var/lib/docker/overlay2/958ae73a3ae371bcd406adcbaf5082bfaf9a750f75b451f1afe78de0d00ed8c2/merged",
                "UpperDir": "/var/lib/docker/overlay2/958ae73a3ae371bcd406adcbaf5082bfaf9a750f75b451f1afe78de0d00ed8c2/diff",
                "WorkDir": "/var/lib/docker/overlay2/958ae73a3ae371bcd406adcbaf5082bfaf9a750f75b451f1afe78de0d00ed8c2/work"
            },
            "Name": "overlay2"
        },
        "Mounts": [
            {
                "Type": "bind",
                "Source": "/users/mag//models/server",
                "Destination": "/models",
                "Mode": "",
                "RW": true,
                "Propagation": "rprivate"
            }
        ],
        "Config": {
            "Hostname": "30aa241b7865",
            "Domainname": "",
            "User": "501:20",
            "AttachStdin": false,
            "AttachStdout": true,
            "AttachStderr": true,
            "ExposedPorts": {
                "9000/tcp": {},
                "9001/tcp": {}
            },
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                "INSTALL_RPMS_FROM_URL=http://repository.toolbox.iotg.sclab.intel.com/ovms-deps/centos-7.8.2003-ovms-deps-rpms.tar.xz",
                "HDDL_INSTALL_DIR=/ovms/lib/hddl",
                "GPU=0"
            ],
            "Cmd": [
                "--model_path",
                "/models/face-detection",
                "--model_name",
                "face-detection",
                "--port",
                "9000",
                "--rest_port",
                "9001",
                "--log_level",
                "DEBUG"
            ],
            "Image": "openvino/model_server:latest",
            "Volumes": null,
            "WorkingDir": "/",
            "Entrypoint": [
                "/ovms/bin/ovms"
            ],
            "OnBuild": null,
            "Labels": {
                "org.label-schema.build-date": "20200809",
                "org.label-schema.license": "GPLv2",
                "org.label-schema.name": "CentOS Base Image",
                "org.label-schema.schema-version": "1.0",
                "org.label-schema.vendor": "CentOS",
                "org.opencontainers.image.created": "2020-08-09 00:00:00+01:00",
                "org.opencontainers.image.licenses": "GPL-2.0-only",
                "org.opencontainers.image.title": "CentOS Base Image",
                "org.opencontainers.image.vendor": "CentOS"
            }
        },
        "NetworkSettings": {
            "Bridge": "",
            "SandboxID": "53e4f967fad224fab556fb535624a4d8d75a046b58015a5c3568f326808b5334",
            "HairpinMode": false,
            "LinkLocalIPv6Address": "",
            "LinkLocalIPv6PrefixLen": 0,
            "Ports": {
                "9000/tcp": [
                    {
                        "HostIp": "0.0.0.0",
                        "HostPort": "9000"
                    }
                ],
                "9001/tcp": [
                    {
                        "HostIp": "0.0.0.0",
                        "HostPort": "9001"
                    }
                ]
            },
            "SandboxKey": "/var/run/docker/netns/53e4f967fad2",
            "SecondaryIPAddresses": null,
            "SecondaryIPv6Addresses": null,
            "EndpointID": "b6f67be614aec99f296444e0b615cd2b071c1c4607fee61ebe6c68465391a512",
            "Gateway": "172.17.0.1",
            "GlobalIPv6Address": "",
            "GlobalIPv6PrefixLen": 0,
            "IPAddress": "172.17.0.2",
            "IPPrefixLen": 16,
            "IPv6Gateway": "",
            "MacAddress": "02:42:ac:11:00:02",
            "Networks": {
                "bridge": {
                    "IPAMConfig": null,
                    "Links": null,
                    "Aliases": null,
                    "NetworkID": "e1aa7fe7df3fe531ba76d485fd00b29e89818025da7a5b46da596068ae52ab24",
                    "EndpointID": "b6f67be614aec99f296444e0b615cd2b071c1c4607fee61ebe6c68465391a512",
                    "Gateway": "172.17.0.1",
                    "IPAddress": "172.17.0.2",
                    "IPPrefixLen": 16,
                    "IPv6Gateway": "",
                    "GlobalIPv6Address": "",
                    "GlobalIPv6PrefixLen": 0,
                    "MacAddress": "02:42:ac:11:00:02",
                    "DriverOpts": null
                }
            }
        }
    }
]

```

Is any other way to verify all required components are running?

Thanks for all your help.

Marcelo

**Comment by dtrawins** [comment]
@magallardo That is very weird. What is the CPU platform and your MAC model? Are you getting the same symptoms on other machines?
We documented how to debug OVMS code in the [developer guide](https://github.com/openvinotoolkit/model_server/blob/main/docs/developer_guide.md#debugging-) I'm not able to reproduce such behaviour on my Macbook Pro. Perhaps this debug procedure could reveal the root cause.

**Comment by magallardo** [comment]
@dtrawins Just to give you some more information, I am running the Openvino Model Server on a MacBook Pro with 2.6 GHz Quad-Core Intel Core i7.  The OS is MacOS Catalina 10.15.6.

Also, I have installed Openvino framework and all the samples including the Json samples work ok.  I am only having issues running the docker container for Openvino Model Server.

Thanks,
Marcelo

**Comment by dtrawins** [comment]
@magallardo Do you observe the same symptoms with the latest docker image `openvino/model_server:2021.2`. 
I was not able to reproduce those symptoms on MAC hosts. A possible troubleshooting could be via starting ovms with [strace](https://man7.org/linux/man-pages/man1/strace.1.html) or running the ovms via [gdb](docs/developer_guide.md#debug). To test it with strace, you would need to start the docker container with updated entrypoint `--entrypoint bash` and `--user root`, installed strace with`yum install strace` and starting ovms like `strace /ovms/bin/ovms --model_name ...`

**Comment by rasapala** [comment]
@magallardo I assume the issue is resolved with the latest 2021.2  ovms image ? Will close the bug if no reply in 2 weeks.
